dask/dataframe/shuffle.py

```diff
@@ -791,11 +791,8 @@ def set_partitions_pre(s, divisions, ascending=True, na_position="last"):
         else:
             partitions = len(divisions) - divisions.searchsorted(s, side="right") - 1
     except TypeError:
-        # When `searchsorted` fails with `TypeError`, it may be
-        # caused by nulls in `s`. Try again with the null-values
-        # explicitly mapped to the first partition.
+        # `searchsorted` fails if `s` contains nulls and strings
```

Comment:
Do we know if the handling in this `except` block is particularly expensive (i.e computation of `s.notna` and indexing of `partitions` and `s`)? Since we are already computing and indexing over `s.isna` always, I am wondering if we should just execute the contents of this block by default.

Reply:
I'm not sure if this is expensive.  I would defer to @rjzamora who may be better informed.

Reply:
I'm pretty sure that the null handling case was implemented in an `except` code path because we were explicitly optimizing at the time for a case where there were no nulls.  I'm not completely sure, but I wouldn't expect there to be much overhead here.

I guess there is some overhead in the temporary `empty` array creation. Perhaps it would be more efficient to use `zeros`? My understanding is that `zeros` will not even reserve/allocate memory for elements that are not assigned (but empty will).

Reply:
After chatting with @quasiben internally, I think that maybe the best option for now would be to merge this PR and resolve dask-cudf test failures, and then open a follow up addressing potential optimizations for `set_partitions_pre`; happy to ping you on that when I open it 🙂

Reply:
Yes - 100% Agree with that plan.